runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java

```diff
@@ -203,6 +232,39 @@ private boolean checkCoreArchitecture() {
         return true;
     }
 
+    private boolean verifyPackageInfo(PackageInfo packageInfo,
+            String algorithm, String fingerprint) {
+        if (packageInfo.signatures == null) {
+            Log.d(TAG, "No signature in package info");
```

Comment:
Sorry, my misunderstand :-) I'll change the log level to `Log.e`.

We do not just publish the APK of Crosswalk runtime library on Google Play Store, but also on our official website, because a lot of devices are isolated from Google service. The developer could self-deploy their own runtime library and specify the download URL, so the user could get the specific APK file. For this case, the next thing we plan to do is to provide the feasibility that the developers could customize the package name of the runtime library, as well as the hashing algorithm and fingerprint (as you see, I extracted these two variables from `verifyPackageInfo()` as parameter) in order to avoid conflict with official package.

Let's assume that two runtime library from different developers happen to have the same package name, but apparently they do not likely have the same signature. So if the signature verification failed, it just means the runtime library installed on device is not expected to be used by current application, but it's not malicious.

Anyway, I have to admit that it is a mess, and we should document this clearly on website to avoid this kind of situation once this feature is implemented. The error message should also be refined in the future.